File: Cores/OSOL.Extremum.Core.DotNet/CodeFeatures/FunctionParser/TreeFunctions/IntervalTreeFunction.cs
using OSOL.Extremum.Core.DotNet.Arithmetics;
using OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser.Trees;
using OSOL.Extremum.Core.DotNet.Vectors;

namespace OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser.TreeFunctions
{
    public class IntervalTreeFunction
    {
        public Tree<Interval> Tree;
        
        public IntervalTreeFunction(Tree<Interval> tree)
        {
            this.Tree = tree;
        }

        public Interval Calculate(IntervalVector v) => Tree.Calculate(v);
    }
}

File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/CodeFeatures/FunctionParser/TreeFunctions/IntervalTreeFunction.scala
package OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.TreeFunctions

import OSOL.Extremum.Core.Scala.Arithmetics.Interval
import OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.Parser
import OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.Trees.Tree
import OSOL.Extremum.Core.Scala.Vectors.IntervalVector

class IntervalTreeFunction(t: Tree[Interval]) {
  def apply(v: IntervalVector): Interval = t.calculate(v)
}

object IntervalTreeFunction {
  def apply(str: String): IntervalTreeFunction = new IntervalTreeFunction(Parser.parseToIntervalTree(str))
}